core/src/main/scala/org/apache/spark/scheduler/EventLogFileWriters.scala

```diff
@@ -0,0 +1,444 @@
+
+  private val logDirForAppPath = getAppEventLogDirPath(logBaseDir, appId, appAttemptId)
+
+  // Only defined if the file system scheme is not local
+  private var hadoopDataStream: Option[FSDataOutputStream] = None
+  private var countingOutputStream: Option[CountingOutputStream] = None
+  private var writer: Option[PrintWriter] = None
+
+  // seq and event log path will be updated soon in rollNewEventLogFile, which `start` will call
+  private var sequence: Long = 0L
+  private var currentEventLogFilePath: Path = logDirForAppPath
+
+  override def start(): Unit = {
+    requireLogBaseDirAsDirectory()
+
+    if (fileSystem.exists(logDirForAppPath) && shouldOverwrite) {
+      // try to delete the directory
+      fileSystem.delete(logDirForAppPath, true)
+    }
+
+    if (fileSystem.exists(logDirForAppPath)) {
+      // we tried to delete the existing one, but failed
+      throw new IOException(s"Target log directory already exists ($logDirForAppPath)")
+    } else {
+      fileSystem.mkdirs(logDirForAppPath, EventLogFileWriter.LOG_FILE_PERMISSIONS)
+      createAppStatusFile(inProgress = true)
+      rollNewEventLogFile(None)
+    }
+  }
+
+  override def writeEvent(eventJson: String, flushLogger: Boolean = false): Unit = {
+    writer.foreach { w =>
+      val currentLen = countingOutputStream.get.getBytesWritten
+      if (currentLen + eventJson.length > eventFileMaxLengthKiB * 1024) {
+        rollNewEventLogFile(Some(w))
+      }
+    }
+
+    // if the event log file is rolled over, writer will refer next event log file
+
+    // scalastyle:off println
+    writer.foreach(_.println(eventJson))
+    // scalastyle:on println
+    if (flushLogger) {
+      writer.foreach(_.flush())
+      hadoopDataStream.foreach(_.hflush())
+    }
+  }
+
+  private def rollNewEventLogFile(w: Option[PrintWriter]): Unit = {
+    writer.foreach(_.close())
+
+    sequence += 1
+    currentEventLogFilePath = getEventLogFilePath(logDirForAppPath, appId, appAttemptId, sequence,
+      compressionCodecName)
+
+    val streams = initLogFile(currentEventLogFilePath)
+    hadoopDataStream = streams._1
+    countingOutputStream = streams._2
+    if (countingOutputStream.isDefined) {
+      writer = Some(new PrintWriter(streams._2.get))
+    } else {
+      writer = None
+    }
+  }
+
+  override def stop(): Unit = {
+    writer.foreach(_.close())
+    val appStatusPathIncomplete = getAppStatusFilePath(logDirForAppPath, appId, appAttemptId,
+      inProgress = true)
+    val appStatusPathComplete = getAppStatusFilePath(logDirForAppPath, appId, appAttemptId,
+      inProgress = false)
+    renameFile(appStatusPathIncomplete, appStatusPathComplete, overwrite = true)
+  }
+
+  override def logPath: String = logDirForAppPath.toString
+
+  private def createAppStatusFile(inProgress: Boolean): Unit = {
+    val appStatusPath = getAppStatusFilePath(logDirForAppPath, appId, appAttemptId, inProgress)
+    val streams = initLogFile(appStatusPath)
```

Comment:
Is there any point in calling `initLogFile` here instead of just `fs.create()`?

Reply:
You're right. Will change.